File: usecases/src/main/java/com/zeyad/usecases/network/ProgressInterceptor.java
package com.zeyad.usecases.network;

import android.support.annotation.NonNull;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Response;

/**
 * @author by ZIaDo on 7/11/17.
 */

public abstract class ProgressInterceptor implements Interceptor {

    private ProgressListener progressListener;

    protected ProgressInterceptor(ProgressListener progressListener) {
        this.progressListener = progressListener;
    }

    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {
        Response originalResponse = chain.proceed(chain.request());
        Response.Builder responseBuilder = originalResponse.newBuilder();
        if (isUpDownload(originalResponse)) {
            responseBuilder.body(new ProgressResponseBody(originalResponse.body(), progressListener));
        } else {
            responseBuilder.body(originalResponse.body());
        }
        return responseBuilder.build();
    }

    public abstract boolean isUpDownload(Response originalResponse);

}

File: usecases/src/main/java/com/zeyad/usecases/network/ProgressInterceptor.kt
package com.zeyad.usecases.network

import okhttp3.Interceptor
import okhttp3.Response

/**
 * @author by ZIaDo.
 */
abstract class ProgressInterceptor(private val progressListener: ProgressListener) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val originalResponse = chain.proceed(chain.request())
        val responseBuilder = originalResponse.newBuilder()
        if (isFileIO(originalResponse)) {
            responseBuilder.body(ProgressResponseBody(originalResponse.body(), progressListener))
        } else {
            responseBuilder.body(originalResponse.body())
        }
        return responseBuilder.build()
    }

    abstract fun isFileIO(originalResponse: Response): Boolean
}
